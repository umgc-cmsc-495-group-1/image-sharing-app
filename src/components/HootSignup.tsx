import React, {useState} from "react";
import {
  Avatar,
  Box,
  Button,
  Container,
  Grid,
  Link,
  TextField,
  Typography,
  Checkbox,
  FormControlLabel,
} from "@mui/material";
import LockOutlinedIcon from "@mui/icons-material/LockOutlined";
import {signInGooglePopup, signup} from "../data/authFunctions"; // , signup
import { UserInterface } from "../types/authentication";
import ErrorsDisplay from "./ErrorsDisplay";
import {useNavigate} from "react-router-dom";
import {uploadProfileImg} from "../data/photoData";

export default function HootSignup() {
  const [createdUser, setCreatedUser] = useState<UserInterface>({
    displayName: "",
    photoURL: "",
    email: "",
    isVerified: false,
    password: "",
    verifyPassword: "",
  });
  const [profileImage, setProfileImage] = useState<string>("");
  const [fileToUpload, setFileToUpload] = useState<File | undefined>(undefined);
  const [errors, setErrors] = useState<string[]>([]);
  const [registerDisabled, setRegisterDisabled] = useState<boolean>(true);
  const EMAIL_REGEX = /^(([^<>()[\]\\.,;!!#$%&*:\s@"]+(\.[^<>()[\]\\.,;!#$%&*:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  const ILLEGAL_CHARACTERS_REGEX = /\W/gi;
  const navigate = useNavigate();
  const FormControlLabelText: React.ReactNode = (
    <Typography variant="body2" color="textSecondary">
      By signing up, you agree to our{" "}
      <Link href="/terms-of-service" color="inherit">
        Terms of Service
      </Link>{" "}
      and{" "}
      <Link href="/privacy" color="inherit">
        Privacy Policy
      </Link>
    </Typography>
  );

  const uploadProfileImage = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { files } = e.target;
    if (files !== null && files.length > 0) {
      const url = URL.createObjectURL(files[0]);
      setFileToUpload(files[0]);
      setProfileImage(url);
    } else {
      setFileToUpload(undefined);
      setProfileImage("");
    }
  };

  const sanitizeDisplayName = (displayName: string) => {
    return displayName.replace(ILLEGAL_CHARACTERS_REGEX, "");
  };

  const checkEmptyValues = () => {
    setErrors([]);
    if (
      createdUser.displayName === "" ||
      createdUser.email === "" ||
      createdUser.password === "" ||
      createdUser.verifyPassword === ""
    ) {
      setErrors((errors) => [...errors, "All fields are required!"]);
      return false;
    }
    if (createdUser.password !== createdUser.verifyPassword) {
      setErrors((errors) => [...errors, "Passwords do not match!"]);
      return false;
    }
    console.log(createdUser.displayName.length)
    console.log(createdUser.displayName)
    if (createdUser.displayName.length < 6 || createdUser.displayName.length > 20) {
      setErrors((errors) => [...errors, "Display name must be between 6 and 20 characters!"]);
      return false;
    }
    if (!EMAIL_REGEX.test(createdUser.email)) {
      setErrors((errors) => [...errors, "Email is not valid!"]);
      return false;
    }
    if (fileToUpload === undefined) {
      setErrors((errors) => [...errors, "Please upload a profile image!"]);
      return false;
    }

    return true;
  }


  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    console.log(event)
    event.preventDefault();
    const user: UserInterface = {
      displayName: createdUser.displayName,
      photoURL: profileImage,
      email: createdUser.email,
      isVerified: false,
      password: createdUser.password,
      verifyPassword: createdUser.verifyPassword,
    };
    // prevent form submission if there are errors
    if (!checkEmptyValues()) return;
    console.log(user)

    try {
      // sign up the user
      await signup(user)
        .then((res) => {
          if (res !== undefined){
            uploadProfileImg(res.user, fileToUpload);
            navigate("/explore");
          }
        })
        .catch((err) => {
          console.error([err.code] + `: ${err.message}`);
        });
    } catch (error) {
      console.error(error);
    }

  };

  const handleGoogleSignin = async () => {
    await signInGooglePopup()
      .then(() => {
        navigate("/explore");
      })
      .catch((err) => {
        console.error(err);
      });
  };

  return (
    <Container component="main" maxWidth="xs">
      <Box
        sx={{
          marginTop: 8,
          display: "flex",
          flexDirection: "column",
          alignItems: "center",
        }}
      >
        <Avatar sx={{ m: 1, bgcolor: "secondary.main" }}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography>Sign Up</Typography>
        <Box
          component="form"
          noValidate
          onSubmit={handleSubmit}
          sx={{ mt: 3 }}
          role="signup-form"
        >
          <Grid container spacing={2}>
            <Grid item xs={12}>
              <Box>
                <ErrorsDisplay errors={errors} />
              </Box>
            </Grid>
            <Grid item xs={12}>
                <Button
                  fullWidth
                  variant="contained"
                  sx={{
                    mt: 3, mb: 2
                }}
                >
                  <label htmlFor="profile-image">Please select a profile image</label>
                </Button>
                <TextField
                  sx={{
                    display: "none",
                  }}
                  required
                  fullWidth
                  id="profile-image"
                  type="file"
                  onChange={uploadProfileImage}
                  inputProps={{
                    accept: "image/*",
                    id: "profile-image",
                    placeholder: "Select Profile Image",
                    style: { display: "none" },
                  }}
                />
            </Grid>
            <Grid item xs={12}>
              <TextField
                onChange={(event) => {
                  setCreatedUser({
                    ...createdUser,
                    displayName: sanitizeDisplayName(event.target.value)
                  });
                }}
                value={createdUser.displayName}
                required
                fullWidth
                id="displayName"
                label="Display Name"
                name="displayName"
                autoComplete="display-name"
                role="display-name-input"
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                onChange={(event) => {
                  setCreatedUser({
                    ...createdUser,
                    email: event.target.value
                  });
                }}
                value={createdUser.email}
                required
                fullWidth
                id="email"
                label="Email Address"
                name="email"
                autoComplete="email"
                role="email-input"
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                onChange={(event) => {
                  setCreatedUser({
                    ...createdUser,
                    password: event.target.value
                  });
                }}
                value={createdUser.password}
                required
                fullWidth
                name="password"
                label="Password"
                type="password"
                id="password"
                autoComplete="new-password"
                role="password-input"
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                onChange={(event) => {
                  setCreatedUser({
                    ...createdUser,
                    verifyPassword: event.target.value
                  });
                }}
                value={createdUser.verifyPassword}
                required
                fullWidth
                name="verifyPassword"
                label="Verify Password"
                type="password"
                id="verifyPassword"
                autoComplete="new-password"
                role="verify-password-input"
              />
            </Grid>
            <Grid item xs={12}>
              <FormControlLabel
                control={
                  <Checkbox
                    checked={!registerDisabled}
                    onChange={(event) => {
                      setRegisterDisabled(!event.target.checked);
                    }}
                    inputProps={{"aria-label": "submit-checkbox-agreement"}}
                  />
                }
                label={FormControlLabelText}
                labelPlacement='end'
              />
            </Grid>
          </Grid>
          <Button
            type="submit"
            fullWidth
            disabled={registerDisabled}
            variant="contained"
            sx={{ mt: 3, mb: 2 }}
          >
            Sign Up
          </Button>
          <Typography textAlign="center"> - or -</Typography>
          <Button
            fullWidth
            disabled={registerDisabled}
            variant="contained"
            sx={{ mt: 3, mb: 2 }}
            onClick={handleGoogleSignin}
          >
            Sign In with Google!
          </Button>
          <Grid container justifyContent="flex-end">
            <Grid item>
              <Link href="/login" variant="body2">
                Already have an account? Log in!
              </Link>
            </Grid>
          </Grid>
        </Box>
      </Box>
    </Container>
  );
}
