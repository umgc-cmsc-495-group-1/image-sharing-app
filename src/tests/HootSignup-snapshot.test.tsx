import React from 'react';
import "@testing-library/react/dont-cleanup-after-each";
import '@testing-library/jest-dom';
import {
  MemoryRouter,
  Routes,
  Route
} from 'react-router-dom';
import HootSignup from '../components/HootSignup';
import { create } from 'react-test-renderer';

function renderWithMemoryRouter(component: JSX.Element) {
  return create(
    <MemoryRouter initialEntries={['/signup']}>
      <Routes>
        <Route path='/signup' element={component} />
      </Routes>
    </MemoryRouter>
  );
}

it('Renders correctly', () => {
  const renderer = renderWithMemoryRouter(<HootSignup />);
  expect(renderer.toJSON()).toMatchSnapshot();
});