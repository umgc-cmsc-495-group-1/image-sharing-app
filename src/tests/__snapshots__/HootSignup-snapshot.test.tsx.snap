// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<main
  className="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
>
  <div
    className="MuiBox-root css-binzgt"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1999axt-MuiAvatar-root"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="LockOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
        />
      </svg>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      Sign Up
    </p>
    <form
      className="MuiBox-root css-178yklu"
      data-testid="SignupTest"
      noValidate={true}
      onSubmit={[Function]}
      role="signup-form"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor="displayName"
              id="displayName-label"
            >
              Display Name
              <span
                aria-hidden={true}
                className="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="name"
                autoFocus={true}
                className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="displayName"
                name="displayName"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-1ftyaf0"
                >
                  <span>
                    Display Name
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor="username"
              id="username-label"
            >
              User Name
              <span
                aria-hidden={true}
                className="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="username"
                autoFocus={false}
                className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="username"
                name="username"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-1ftyaf0"
                >
                  <span>
                    User Name
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor="email"
              id="email-label"
            >
              Email Address
              <span
                aria-hidden={true}
                className="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="email"
                autoFocus={false}
                className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="email"
                name="email"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-1ftyaf0"
                >
                  <span>
                    Email Address
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor="password"
              id="password-label"
            >
              Password
              <span
                aria-hidden={true}
                className="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="new-password"
                autoFocus={false}
                className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="password"
                name="password"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="password"
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-1ftyaf0"
                >
                  <span>
                    Password
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor="verifyPassword"
              id="verifyPassword-label"
            >
              Verify Password
              <span
                aria-hidden={true}
                className="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="new-password"
                autoFocus={false}
                className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="verifyPassword"
                name="verifyPassword"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="password"
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-1ftyaf0"
                >
                  <span>
                    Verify Password
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          
        </div>
      </div>
      <button
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root Mui-disabled  css-1vhaqj4-MuiButtonBase-root-MuiButton-root"
        disabled={true}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="submit"
      >
        Sign Up
      </button>
      <button
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-1vhaqj4-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Google Sign In
      </button>
      <div
        className="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <a
            className="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineAlways css-qk1173-MuiTypography-root-MuiLink-root"
            href="/login"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            Already have an account? Log in!
          </a>
        </div>
      </div>
    </form>
  </div>
</main>
`;
